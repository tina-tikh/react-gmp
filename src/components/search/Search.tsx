import * as React from 'react';
import { Component, ReactNode } from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';

import { AppState, getSearchMovies, getSearchTotal } from '../../store';
import { thunkReceiveMovies } from '../../thunks';
import { Movie, SearchBy, SortBy } from '../../api';
import { ActionBar, ActionBarCaption, Header, Main, MainMessage } from '../layout';
import { ToggleGroup } from '../form';
import FilmList from '../FilmList';
import SearchBox from './SearchBox';
import SearchCaption from './SearchCaption';
import SearchHeader from './SearchHeader';
import SearchManager from './SearchManager';

interface SearchProps extends RouteComponentProps<any> {
  movies: Movie[],
  total: number,
  thunkReceiveMovies: typeof thunkReceiveMovies
}

class Search extends Component<SearchProps> {
  private searchManager: SearchManager<any>;

  constructor(props: SearchProps) {
    super(props);
    this.searchManager = new SearchManager(props.location);
  }

  componentDidUpdate(prevProps: Readonly<SearchProps>) {
    if (this.props.location !== prevProps.location) {
      const params = this.searchManager.getSearchParamsObj();
      this.props.thunkReceiveMovies(params);
    }
  }

  render(): ReactNode {
    const params = this.searchManager.getSearchParamsObj();
    const total = this.props.total;
    const movies = this.props.movies;

    return (
      <React.Fragment>
        <Header>
          <SearchHeader>
            <SearchCaption>Find your movie</SearchCaption>
            <SearchBox searchValue={params.search}
                       onSubmit={(e: string) => this.handleSubmit(e)}
            ></SearchBox>
            <ToggleGroup
              label="Search by"
              value={params.searchBy}
              values={Object.values(SearchBy)}
              valueLabels={Object.keys(SearchBy)}
              onChange={(e: string) => this.handleSearchByChange(e)}
            ></ToggleGroup>
          </SearchHeader>
        </Header>
        <Main>
          <ActionBar>
            <ActionBarCaption>{total} movie found</ActionBarCaption>
            <ToggleGroup
              label="Sort by"
              value={params.sortBy}
              values={Object.values(SortBy)}
              valueLabels={Object.keys(SortBy)}
              onChange={(e: string) => this.handleSortByChange(e)}
            ></ToggleGroup>
          </ActionBar>
          <FilmList films={movies}/>
          {total === 0 && <MainMessage>No films found</MainMessage>}
        </Main>
      </React.Fragment>
    );
  }

  handleSearchByChange(searchByValue: string): void {
    this.searchManager.setSearchBy(searchByValue as SearchBy);
    this.refresh();
  }

  handleSortByChange(sortByValue: string): void {
    this.searchManager.setSortBy(sortByValue as SortBy);
    this.refresh();
  }

  handleSubmit(q: string): void {
    this.searchManager.setQuery(q);
    this.refresh();
  }

  refresh(): void {
    this.props.history.push('/search?' + this.searchManager.queryParams);
  }
}

const mapStateToProps = (state: AppState) => ({
  movies: getSearchMovies(state),
  total: getSearchTotal(state)
});

export default connect(
  mapStateToProps,
  { thunkReceiveMovies }
)(Search);
