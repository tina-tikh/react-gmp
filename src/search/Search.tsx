import * as React from 'react';

import { ActionBar, ActionBarCaption, FilmList, Header, Main, MainMessage, Movie, ToggleGroup } from '../common';
import movies from '../common/api/movies-mock.json';
import SearchBox from './SearchBox';
import SearchCaption from './SearchCaption';
import SearchHeader from './SearchHeader';

type SearchProps = unknown;
type SearchState = {
  searchByValue: string;
  sortByValue: string;
  searchResult: { data: Movie[]; total: number };
};

class Search extends React.Component<SearchProps, SearchState> {
  searchByOpts: string[] = ['Title', 'Genre'];
  sortByOpts: string[] = ['Release', 'Rating'];

  constructor(props: any) {
    super(props);
    this.state = {
      searchByValue: this.searchByOpts[0],
      sortByValue: this.sortByOpts[0],
      searchResult: movies,
    };
  }

  render() {
    const res = this.state.searchResult;
    return (
      <React.Fragment>
        <Header>
          <SearchHeader>
            <SearchCaption>Find your movie</SearchCaption>
            <SearchBox
              onSubmit={(e: string) => this.handleSubmit(e)}
            ></SearchBox>
            <ToggleGroup
              label="Search by"
              value={this.state.searchByValue}
              opts={this.searchByOpts}
              onChange={(e: string) => this.handleSearchByChange(e)}
            ></ToggleGroup>
          </SearchHeader>
        </Header>
        <Main>
          <ActionBar>
            <ActionBarCaption>{res.total} movie found</ActionBarCaption>
            <ToggleGroup
              label="Sort by"
              value={this.state.sortByValue}
              opts={this.sortByOpts}
              onChange={(e: string) => this.handleSortByChange(e)}
            ></ToggleGroup>
          </ActionBar>
          <FilmList films={res.data} />
          {res.total === 0 && <MainMessage>No films found</MainMessage>}
        </Main>
      </React.Fragment>
    );
  }

  handleSearchByChange(searchByValue: string) {
    this.setState({
      searchByValue,
    });
  }

  handleSortByChange(sortByValue: string) {
    this.setState({
      sortByValue,
    });
  }

  handleSubmit(e: string) {
    this.setState({
      searchResult: {
        data: [],
        total: 0,
      },
    });
  }
}

export default Search;
