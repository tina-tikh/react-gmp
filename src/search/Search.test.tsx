import * as React from 'react';
import { shallow } from 'enzyme';

import Search from './Search';

describe('<Search />', () => {
  it('should be rendered', () => {
    const component = shallow(<Search />);
    expect(component).toMatchSnapshot();
  });
});
