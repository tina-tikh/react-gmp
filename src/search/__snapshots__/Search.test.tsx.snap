// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should be rendered 1`] = `
<Fragment>
  <Header>
    <Styled(div)>
      <Styled(h1)>
        Find your movie
      </Styled(h1)>
      <SearchBox
        onSubmit={[Function]}
      />
      <ToggleGroup
        label="Search by"
        onChange={[Function]}
        opts={
          Array [
            "Title",
            "Genre",
          ]
        }
        value="Title"
      />
    </Styled(div)>
  </Header>
  <Styled(main)>
    <Styled(section)>
      <Styled(span)>
         movie found
      </Styled(span)>
      <ToggleGroup
        label="Sort by"
        onChange={[Function]}
        opts={
          Array [
            "Release",
            "Rating",
          ]
        }
        value="Release"
      />
    </Styled(section)>
    <FilmList />
  </Styled(main)>
</Fragment>
`;
